submariner/multiple-cable-drivers.md

```diff
@@ -0,0 +1,192 @@
+spec:
+  leftClusterSelector:
+    database-role: server
+  rightClusterSelector:
+    database-role: client
+```
+
+- We want any cluster to access a database server in another cluster but no other clusters to connect. We label the server
+`Cluster` appropriately and define a single policy with the corresponding label selector and an empty selector to match all.
+
+```yaml
+apiVersion: v1
+kind: ClusterConnectivityPolicy
+metadata:
+  name: client-server-policy
+spec:
+  leftClusterSelector:
+    database-server: true
+  rightClusterSelector:
+```
+
+### Multiple cable drivers API
+
+This following CRD is proposed:
+
+```Go
+type ClusterConnectionPolicies struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec ClusterConnectionPoliciesSpec `json:"spec"`
+}
+
+type ClusterConnectionPoliciesSpec struct {
+    Policies []ClusterConnectionPolicy `json:"policies,omitempty"`
+}
+
+type ClusterConnectionPolicy struct {
+    // LeftClusterSelector identifies the Cluster resources representing the clusters on one end of a connection. An empty
+    // selector indicates wildcard, ie matches any cluster.
+    LeftClusterSelector metav1.LabelSelector `json:"leftClusterSelector,omitempty"`
+
+    // RightClusterSelector identifies the Cluster resources representing the clusters on the other end of a connection.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`
+
+    CableDriver *CableDriverSpec `json:"cableDriver,omitempty"`
+}
+
+type CableDriverSpec struct {
+    // Name of the cable driver implementation.
+    Name string `json:"name"`
+
+    // Options specifies key/value configuration parameters for the cable driver.
+    Options map[string]string `json:"options,omitempty"`
+}
+```
+
+The `ClusterConnectionPolicies` CR defines a list of policies that are applied in order of precedence to select the cable driver
+and associated optional configuration options to use for each cluster connection. To resolve the connection policy for a remote
+cluster's `Endpoint`, Submariner would iterate the list of policies to search for a policy whose left and right label selectors
+match the corresponding `Cluster` resources of the local and remote `Endpoints`. The first matching policy is used. If no matching
+policy is found, the default cable driver is used.
+
+Defining the policies as an ordered list allows the user to resolve ambiguity should a cluster pair match multiple policies.
+Otherwise, if Submariner encountered such a case, it would either have to arbitrarily pick a policy or reject all of them and
+use the default. Neither option is ideal. An alternative to the ordered list would be to have separate resources for each policy
+and define a priority or precedence field. However it seems a single resource without the use of magic priority numbers would be
+easier for the user to maintain.
```

Comment:
I disagree with this premise, using weights won't have any tangible disadvantage versus using a singleton resource.

Using the ordered list doesn't give any better clarity as to which policies affect what since you're lacking the selector part when viewing it, and as an admin would have to query the individual policies for that data anyhow.